crates/matrix-sdk/src/authentication/qrcode/messages.rs

```diff
@@ -0,0 +1,326 @@
+        assert_let!(QrAuthMessage::LoginProtocolAccepted { .. } = &message);
+        let serialized = serde_json::to_value(&message).unwrap();
+        assert_eq!(json, serialized);
+    }
+
+    #[test]
+    fn test_login_success() {
+        let json = json!({
+            "type": "m.login.success",
+        });
+
+        let message: QrAuthMessage = serde_json::from_value(json.clone()).unwrap();
+        assert_let!(QrAuthMessage::LoginSuccess { .. } = &message);
+        let serialized = serde_json::to_value(&message).unwrap();
+        assert_eq!(json, serialized);
+    }
+
+    #[test]
+    fn test_login_declined() {
+        let json = json!({
+            "type": "m.login.declined",
+        });
+
+        let message: QrAuthMessage = serde_json::from_value(json.clone()).unwrap();
+        assert_let!(QrAuthMessage::LoginDeclined { .. } = &message);
+        let serialized = serde_json::to_value(&message).unwrap();
+        assert_eq!(json, serialized);
+    }
+
+    #[test]
+    fn test_login_failure() {
+        let json = json!({
+            "type": "m.login.failure",
+            "reason": "unsupported_protocol",
+            "homeserver": "https://matrix-client.matrix.org/"
+        });
+
+        let message: QrAuthMessage = serde_json::from_value(json.clone()).unwrap();
+        assert_let!(QrAuthMessage::LoginFailure { reason, .. } = &message);
+        assert_eq!(reason, &LoginFailureReason::UnsupportedProtocol);
+        let serialized = serde_json::to_value(&message).unwrap();
+        assert_eq!(json, serialized);
+    }
+
+    #[test]
+    fn test_login_secrets() {
+        let json = json!({
+            "type": "m.login.secrets",
+            "cross_signing": {
+                "master_key": "rTtSv67XGS6k/rg6/yTG/m573cyFTPFRqluFhQY+hSw",
+                "self_signing_key": "4jbPt7jh5D2iyM4U+3IDa+WthgJB87IQN1ATdkau+xk",
+                "user_signing_key": "YkFKtkjcsTxF6UAzIIG/l6Nog/G2RigCRfWj3cjNWeM",
+            },
+            "backup": {
+                "algorithm": "m.megolm_backup.v1.curve25519-aes-sha2",
+                "backup_version": "2",
+                "key": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+            },
+        });
+
+        let message: QrAuthMessage = serde_json::from_value(json.clone()).unwrap();
+        assert_let!(
+            QrAuthMessage::LoginSecrets(SecretsBundle { cross_signing, backup }) = &message
+        );
+        assert_eq!(cross_signing.master_key, "rTtSv67XGS6k/rg6/yTG/m573cyFTPFRqluFhQY+hSw");
+        assert_eq!(cross_signing.self_signing_key, "4jbPt7jh5D2iyM4U+3IDa+WthgJB87IQN1ATdkau+xk");
+        assert_eq!(cross_signing.user_signing_key, "YkFKtkjcsTxF6UAzIIG/l6Nog/G2RigCRfWj3cjNWeM");
+
+        assert_let!(Some(BackupSecrets::MegolmBackupV1Curve25519AesSha2(backup)) = backup);
+        assert_eq!(backup.backup_version, "2");
+        assert_eq!(&backup.key.to_base64(), "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
```

Comment:
AAAAAAAaaaaaaa… *boom*”

— “Are you alright?”

Reply:
It's just 32 bytes of zeroes. Not the best key, but usable for a test and produces a funny and well recognized base64 representation.